### PR TITLE
Feat (Channel-Splitting): sets up first skeleton for channel-splitting

### PR DESCRIPTION
**Pull request by fabianandresgrob**
Implementing Channel Splitting using the pre-existing framework of regions. Simple skeleton for the start.

### CODE REVIEW

File: src/brevitas/nn/split_layer.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+from inspect import signature
+
+import torch
+
+INPUT_NAMES = ['input', 'inp', 'query', 'x']
```
Comment (Giuseppe5) [inline src/brevitas/nn/split_layer.py:5]: Isn't this defined elsewhere? Let's keep it in one place

File: src/brevitas/ptq_algorithms/channel_splitting.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+    clip_max_abs = max_abs * clip_threshold
+    n = 2 ** (bit_width - 1) - 1
+    return n / clip_max_abs
+
+
+def _channels_maxabs(module, splits_per_layer, split_input):
+    # works for Conv2d and Linear
+    dim = 1 - int(split_input)
+    if len(module.weight.data.shape) > 1:
+        if isinstance(module, nn.Conv2d):
```
Comment (Giuseppe5) [inline src/brevitas/ptq_algorithms/channel_splitting.py:28]: Support for all Conv and ConvTranspose

File: src/brevitas/ptq_algorithms/channel_splitting.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+    max_abs = weights.abs().max()
+    clip_max_abs = max_abs * clip_threshold
+    n = 2 ** (bit_width - 1) - 1
+    return n / clip_max_abs
+
+
+def _channels_maxabs(module, splits_per_layer, split_input):
+    # works for Conv2d and Linear
+    dim = 1 - int(split_input)
+    if len(module.weight.data.shape) > 1:
```
Comment (Giuseppe5) [inline src/brevitas/ptq_algorithms/channel_splitting.py:27]: ?

File: src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py [hunk 4]
```diff
@@ -332,7 +363,12 @@ def main():
             model,
             equalize_iters=args.graph_eq_iterations,
             equalize_merge_bias=args.graph_eq_merge_bias,
-            merge_bn=not args.calibrate_bn)
+            merge_bn=not args.calibrate_bn,
```
Comment (fabianandresgrob) [inline src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py:366]: can also change this to merge_bn or remove that flag from argparse

File: src/brevitas/graph/quantize.py [hunk 5]
```diff
@@ -263,7 +264,13 @@ def preprocess_for_quantize(
         equalize_merge_bias=True,
         merge_bn=True,
         equalize_bias_shrinkage: str = 'vaiq',
-        equalize_scale_computation: str = 'maxabs'):
+        equalize_scale_computation: str = 'maxabs',
+        channel_splitting=False,
+        channel_splitting_ratio=0.02,
+        channel_splitting_grid_aware=False,
```
Comment (Giuseppe5) [inline src/brevitas/graph/quantize.py:270]: Is this still a thing?

File: src/brevitas/graph/quantize.py [hunk 6]
```diff
@@ -263,7 +264,13 @@ def preprocess_for_quantize(
...
+        channel_splitting_split_input=True,
+        channel_splitting_criterion: str = 'maxabs',
+        channel_splitting_weight_bit_width=8):
```
Comment (Giuseppe5) [inline src/brevitas/graph/quantize.py:273]: Is this still a thing?

File: src/brevitas/ptq_algorithms/channel_splitting.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+from brevitas.graph.base import GraphTransform
+from brevitas.graph.equalize import _batch_norm
+from brevitas.graph.equalize import _channel_maxabs
+from brevitas.graph.equalize import _extract_regions
+from brevitas.graph.equalize import _get_input_axis
+from brevitas.graph.equalize import _get_output_axis
+from brevitas.graph.equalize import Region
+from brevitas.graph.equalize import transpose
+
+_conv = (
```
Comment (Giuseppe5) [inline src/brevitas/ptq_algorithms/channel_splitting.py:18]: What about linear?

File: src/brevitas/ptq_algorithms/channel_splitting.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        sinks: Dict[str, nn.Module],
+        split_criterion: str,
+        split_ratio: float,
+        split_input: bool) -> Dict[nn.Module, List[torch.Tensor]]:
+    """
+    This method computes the channels that will be split based on `split_criterion`.
+    """
+    modules = sinks if split_input else sources
+    _get_axis = _get_input_axis if split_input else _get_output_axis
+    # the modules are all of the same shape so we can just take the first one
```
Comment (Giuseppe5) [inline src/brevitas/ptq_algorithms/channel_splitting.py:33]: Not sure this is always true, e.g. AlexNet has a region where there is a shape mismatch because of a flatten.

File: src/brevitas/ptq_algorithms/channel_splitting.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            max_per_channel = _channel_maxabs(weight_t.reshape(weight_t.size(0), -1))
+            channels_sorted = torch.argsort(max_per_channel, descending=True)
+            module_to_channels[name] = channels_sorted[:splits_per_layer]
+
+    # return tensor with the indices to split
+    channels_to_split = torch.cat(list(module_to_channels.values()))
+    return torch.unique(channels_to_split)
+
+
+def transpose_tensor(tensor: torch.Tensor, axis: int):
```
Comment (Giuseppe5) [inline src/brevitas/ptq_algorithms/channel_splitting.py:55]: Pretty sure something similar it's defined already in equalize.py

File: src/brevitas/ptq_algorithms/channel_splitting.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    shape = list(range(tensor.ndim))
+    axis = shape[axis]
+    shape.insert(0, axis)
+    del shape[axis + 1]
+    return tensor.permute(shape)
+
+
+def _split_channels(module, channels_to_split, split_input=False, split_factor=0.5) -> None:
+    """
+    Splits the channels `channels_to_split` of the `weights`.
```
Comment (Giuseppe5) [inline src/brevitas/ptq_algorithms/channel_splitting.py:69]: The docstring is a bit weird. Maybe rewording.
Also if you add the Return keyword, then we might need to specify the entire function signature. Remove it for now

File: src/brevitas/ptq_algorithms/channel_splitting.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    del orig_shape[axis]
+    weight_t = weight_t.reshape(weight_t.size(0), *orig_shape)
+    weight_t = transpose_tensor(weight_t, axis)
+    module.weight.data = weight_t
+    if isinstance(module, _conv):
+        if split_input:
+            module.in_channels += num_added_channels
+        else:
+            module.out_channels += num_added_channels
+    else:
```
Comment (Giuseppe5) [inline src/brevitas/ptq_algorithms/channel_splitting.py:107]: It's better to avoid a catch-all "Else"

File: src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py (hunk 4 above)
Comment (Giuseppe5) [reply src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py:366]: Let's change it

File: src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py [hunk 12]
```diff
@@ -400,9 +436,9 @@ def main():
             iters=args.learned_round_iters,
             optimizer_lr=args.learned_round_lr)
 
-    if args.calibrate_bn:
-        print("Calibrate BN:")
-        calibrate_bn(calib_loader, quant_model)
+    # if args.calibrate_bn:
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py:439]: Restore

File: tests/brevitas/graph/test_channel_splitting.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+    assert torch.allclose(expected_out, out, atol=ATOL)
+
+
+@pytest.mark.parametrize('split_ratio', [0.05, 0.1, 0.2])
+@pytest.mark.parametrize('split_input', [False, True])
+def test_transpose_layers(split_ratio, split_input):
+    torch.manual_seed(SEED)
+
+    model = torch.nn.Sequential(
+        torch.nn.ConvTranspose2d(in_channels=4, out_channels=8, kernel_size=3),
```
Comment (Giuseppe5) [inline tests/brevitas/graph/test_channel_splitting.py:57]: Do we need 3?

File: tests/brevitas/graph/test_channel_splitting.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+from brevitas.fx import symbolic_trace
+from brevitas.graph.fixed_point import MergeBatchNorm
+from brevitas.ptq_algorithms.channel_splitting import *
+
+from .equalization_fixtures import *
+
+
+@pytest.mark.parametrize('split_ratio', [0.05, 0.1, 0.2])
+@pytest.mark.parametrize('split_input', [False, True])
+def test_resnet18(split_ratio, split_input):
```
Comment (Giuseppe5) [inline tests/brevitas/graph/test_channel_splitting.py:14]: Check the equalization tests, and re-use some of the utilities there.

File: tests/brevitas/graph/test_channel_splitting.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+    model = MergeBatchNorm().apply(model)
+
+    model = RegionwiseChannelSplitting(
+        split_ratio=split_ratio, split_input=split_input).apply(model)
+    out = model(inp)
+    assert torch.allclose(expected_out, out, atol=ATOL)
+
+
+@pytest_cases.fixture
+def convgroupconv_model():
```
Comment (Giuseppe5) [inline tests/brevitas/graph/test_channel_splitting.py:100]: Put this together with the equalization tests.

File: tests/brevitas/graph/test_channel_splitting.py (hunk 14 above)
Comment (Giuseppe5) [reply tests/brevitas/graph/test_channel_splitting.py:14]: Let's have 2 group of tests:

- One for the torchvision models
- One for the toy examples

Mark with `pytest.skip` everything that is not compatible with channel splitting. Modify the tests to merge batch norm only for channel splitting (just because otherwise we would need to update all the other tests).

File: src/brevitas/ptq_algorithms/channel_splitting.py (hunk 8 above)
Comment (fabianandresgrob) [reply src/brevitas/ptq_algorithms/channel_splitting.py:33]: You're right, but at this point a region with that case should be filtered out

File: src/brevitas/graph/channel_splitting.py [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+            max_per_channel = _channel_maxabs(weight_t.reshape(weight_t.size(0), -1))
+            channels_sorted = torch.argsort(max_per_channel, descending=True)
+            module_to_channels[name] = channels_sorted[:splits_per_layer]
+
+    # return tensor with the indices to split
+    channels_to_split = torch.cat(list(module_to_channels.values()))
+    return torch.unique(channels_to_split)
+
+
+def transpose_tensor(tensor: torch.Tensor, axis: int):
```
Comment (Giuseppe5) [inline src/brevitas/graph/channel_splitting.py:59]: No sharing?

File: src/brevitas/graph/channel_splitting.py [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+def _clean_regions(regions: List[Region]):
+    """
+    Given a list of regions, this method removes all regions that are not compatible with channel splitting.
+    """
+    # idea: map modules to their regions and check whether it appears in multiple regions
+    regions_to_del = set()
+    source_modules = dict()
+    sink_modules = dict()
+    for i, region in enumerate(regions):
+        sources = {src: region.get_module_from_name(src) for src in region.srcs_names}
```
Comment (Giuseppe5) [inline src/brevitas/graph/channel_splitting.py:210]: why pointer to modules instead of names?

File: src/brevitas/graph/quantize.py [hunk 18]
```diff
@@ -263,7 +264,11 @@ def preprocess_for_quantize(
         equalize_merge_bias=True,
         merge_bn=True,
         equalize_bias_shrinkage: str = 'vaiq',
-        equalize_scale_computation: str = 'maxabs'):
+        equalize_scale_computation: str = 'maxabs',
+        channel_splitting=False,
```
Comment (Giuseppe5) [inline src/brevitas/graph/quantize.py:268]: Let's use the ratio to define whether this is enabled or not. Channel split = 0 is disabled.
Default ratio I would put disabled since channel splitting does not come for free (unlike weight eq for example)

File: src/brevitas/graph/channel_splitting.py [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+            channels_sorted = torch.argsort(max_per_channel, descending=True)
+            module_to_channels[name] = channels_sorted[:splits_per_layer]
+
+    # return tensor with the indices to split
+    channels_to_split = torch.cat(list(module_to_channels.values()))
+    return torch.unique(channels_to_split)
+
+
+@torch.no_grad()
+def _split_channels(module, channels_to_split, split_input=False, split_factor=0.5) -> None:
```
Comment (Giuseppe5) [inline src/brevitas/graph/channel_splitting.py:60]: typing

File: src/brevitas/graph/channel_splitting.py [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+from brevitas.fx import GraphModule
+from brevitas.graph.base import GraphTransform
+from brevitas.graph.equalize import _channel_maxabs
+from brevitas.graph.equalize import _extract_regions
+from brevitas.graph.equalize import _get_input_axis
+from brevitas.graph.equalize import _get_output_axis
+from brevitas.graph.equalize import Region
+from brevitas.graph.equalize import transpose
+
+__all__ = ['RegionwiseChannelSplitting']
```
Comment (Giuseppe5) [inline src/brevitas/graph/channel_splitting.py:19]: Let's call it GraphChannelSplitting

File: src/brevitas/graph/channel_splitting.py [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+
+    # filter out unsupported layers
+    if any(map(_is_unsupported, sinks + srcs)):
+        return False
+
+    # mha can only be in the sources
+    if any(map(_is_mha, sinks)):
+        return False
+    elif any(map(_is_mha, srcs)):
+        # for all mha in sources, we need to unwrap them access the code directly
```
Comment (Giuseppe5) [inline src/brevitas/graph/channel_splitting.py:163]: comment unclear

File: src/brevitas/graph/channel_splitting.py [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+    if len(sinks_ics) > 1:
+        return False
+
+    return srcs_ocs == sinks_ics
+
+
+def _unwrap_mha(
+    sources: Union[List[nn.Module], Dict[str, nn.Module]]
+) -> Union[List[nn.Module], Dict[str, nn.Module]]:
+    if isinstance(sources, List):
```
Comment (Giuseppe5) [inline src/brevitas/graph/channel_splitting.py:182]: Why list and dict?
Let's try to be consistent and pass it always in the same way. After all, it always comes from a Region structure

File: src/brevitas/graph/quantize.py [hunk 23]
```diff
@@ -285,6 +289,11 @@ def preprocess_for_quantize(
         merge_bias=equalize_merge_bias,
         bias_shrinkage=equalize_bias_shrinkage,
         scale_computation_type=equalize_scale_computation).apply(model)
+    if channel_splitting_ratio:
```
Comment (Giuseppe5) [inline src/brevitas/graph/quantize.py:292]: if channel_splitting_ration >0

File: tests/brevitas/graph/test_channel_splitting.py [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+    # merge BN before applying channel splitting
+    model = MergeBatchNorm().apply(model)
+
+    # save model's state dict to check if channel splitting was done or not
+    old_state_dict = model.state_dict()
+
+    regions = _extract_regions(model)
+    regions = _clean_regions(regions)
+    if not len(regions) > 0:
+        pytest.skip(reason='No regions supported.')
```
Comment (Giuseppe5) [inline tests/brevitas/graph/test_channel_splitting.py:43]: We should check that len(regions) == 0 for only the toymodels we expect it to be zero

File: tests/brevitas/graph/test_channel_splitting.py [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+    model = symbolic_trace(model)
+    # merge BN before applying channel splitting
+    model = MergeBatchNorm().apply(model)
+
+    old_state_dict = model.state_dict()
+
+    regions = _extract_regions(model)
+    regions = _clean_regions(regions)
+    if not len(regions) > 0:
+        pytest.skip(reason='No regions supported.')
```
Comment (Giuseppe5) [inline tests/brevitas/graph/test_channel_splitting.py:87]: Same as before

File: src/brevitas/graph/channel_splitting.py [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+            # add region to be deleted
+            regions_to_del.add(i)
+
+    regions = [regions[i] for i, _ in enumerate(regions) if i not in regions_to_del]
+    return regions
+
+
+class GraphChannelSplitting(GraphTransform):
+
+    def __init__(self, split_ratio=0.02, split_criterion='maxabs', split_input=True):
```
Comment (Giuseppe5) [inline src/brevitas/graph/channel_splitting.py:256]: Typing

File: src/brevitas/graph/channel_splitting.py [hunk 27]
```diff
@@ -0,0 +1,280 @@
...
+        super(GraphChannelSplitting, self).__init__()
+
+        self.split_ratio = split_ratio
+        self.split_criterion = split_criterion
+        self.split_input = split_input
+
+    def apply(
```
Comment (Giuseppe5) [inline src/brevitas/graph/channel_splitting.py:263]: Half of the typing

File: src/brevitas/graph/quantize.py [hunk 28]
```diff
@@ -263,7 +264,10 @@ def preprocess_for_quantize(
         equalize_merge_bias=True,
         merge_bn=True,
         equalize_bias_shrinkage: str = 'vaiq',
-        equalize_scale_computation: str = 'maxabs'):
+        equalize_scale_computation: str = 'maxabs',
+        channel_splitting_ratio=0.0,
+        channel_splitting_split_input=True,
```
Comment (Giuseppe5) [inline src/brevitas/graph/quantize.py:269]: Typing

File: src/brevitas_examples/imagenet_classification/ptq/benchmark/ptq_benchmark_torchvision.py [hunk 29]
```diff
@@ -105,7 +105,10 @@ def unique(sequence):
     'accumulator_bit_width': [16],  # Accumulator bit width, only in combination with GPFA2Q
     'act_quant_percentile': [99.999],  # Activation Quantization Percentile
     'uint_sym_act_for_unsigned_values': [True],  # Whether to use unsigned act quant when possible
-}
+    'channel_splitting_ratio': [0.0
+                               ],  # Channel Splitting ratio, 0.0 means no splitting is performed
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/benchmark/ptq_benchmark_torchvision.py:109]: Can we shorten the comment a bit?

File: src/brevitas/graph/channel_splitting.py [hunk 30]
```diff
@@ -0,0 +1,280 @@
...
+            max_per_channel = _channel_maxabs(weight_t.reshape(weight_t.size(0), -1))
+            channels_sorted = torch.argsort(max_per_channel, descending=True)
+            all_channels.append(channels_sorted[:splits_per_layer])
+
+    # return tensor with the unique indices to split
+    channels_to_split = torch.cat(all_channels)
+    return torch.unique(channels_to_split)
+
+
+@torch.no_grad()
```
Comment (Giuseppe5) [inline src/brevitas/graph/channel_splitting.py:60]: Is this needed?

**Review by Giuseppe5 — APPROVED** [review]
